pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageContainerImpl.java

```diff
@@ -293,14 +301,27 @@ public OpSendMsg createOpSendMsg() throws IOException {
                 messageMetadata.getHighestSequenceId(), numMessagesInBatch, messageMetadata, encryptedPayload);
 
         OpSendMsg op = OpSendMsg.create(messages, cmd, messageMetadata.getSequenceId(),
-                messageMetadata.getHighestSequenceId(), firstCallback);
+                messageMetadata.getHighestSequenceId(), firstCallback, batchAllocatedSize);
 
         op.setNumMessagesInBatch(numMessagesInBatch);
         op.setBatchSizeByte(currentBatchSizeBytes);
         lowestSequenceId = -1L;
         return op;
     }
 
+    private void reserveBatchAllocatedSizeWhenInit(int batchAllocatedInitSize) {
+        batchAllocatedSize = batchAllocatedInitSize;
+        producer.client.getMemoryLimitController().forceReserveMemory(batchAllocatedSize);
+    }
+
+    private void updateBatchAllocatedSizeWhenLeave(int encryptedCapacity) {
+        int delta = encryptedCapacity - batchAllocatedSize;
```

Comment:
This delta variable may be less than 0.

Reply:
It doesn't matter if `delta`  less than 0 or not.  We could see from below that if `size` less than 0,  this method workd well.

https://github.com/apache/pulsar/blob/09f5eeb0c946ee890483e087f802e30a2a2b60ab/pulsar-client/src/main/java/org/apache/pulsar/client/impl/MemoryLimitController.java#L48-L51